packages/jaeger-ui/src/components/TracePage/TraceGraph/OpNode.js

```diff
@@ -0,0 +1,135 @@
+// @flow
+
+// Copyright (c) 2017 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as React from 'react';
+import { Popover, Icon } from 'antd';
+import colorGenerator from '../../../utils/color-generator';
+
+import type { PVertex } from '../../../model/trace-dag/types';
+
+import './OpNode.css';
+
+type Props = {
+  count: number,
+  time: number,
+  percent: number,
+  selfTime: number,
+  operation: string,
+  service: string,
+  registerFn: any,
+};
+type State = {
+  nodeMode: boolean,
+};
+
+function formatPercent(percent) {
+  return Math.round(percent * 100) / 100;
+}
+
+export default class OpNode extends React.PureComponent<Props, State> {
+  props: Props;
+  state: State;
+
+  static helpPopover() {
+    const helpTable = (
+      <table className={`OpNode OpNode--mode-service`}>
+        <tbody>
+          <tr>
+            <td className={`OpNode--metricCell`}>Count</td>
+            <td className={`OpNode--labelCell`}>
+              <strong>Service</strong>
+            </td>
+            <td className={`OpNode--metricCell`}>% of trace duration</td>
+          </tr>
+          <tr>
+            <td className={`OpNode--metricCell`}>Duration</td>
+            <td className={`OpNode--labelCell`}>Operation</td>
+            <td className={`OpNode--metricCell`}>Self time</td>
+          </tr>
+        </tbody>
+      </table>
+    );
+    return (
+      <Popover
+        overlayClassName={`OpNode--popover`}
+        placement="left"
+        mouseEnterDelay={0.25}
+        content={helpTable}
+      >
+        <Icon type="question-circle" />
+      </Popover>
+    );
+  }
+
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      nodeMode: false,
+    };
+    props.registerFn.call(this, this);
+  }
+
+  toggleNodeMode = () => {
```

Comment:
I had previously a similar solution using props. The problem here is when the state is updated in the `TraceGraph` it will execute a whole render of the `DirectedGraph`. This will flicker (minimap is newly loaded and the zoom is reset). So I decided to go this hacky way.

Reply:
To avoid rerendering the `DirectedGraph` on each change to `nodeMode`, `nodeMode` could be a part of the redux state, then each `opNode` would subscribe to this piece of redux state. Since `TraceGraph` is a pure component, it should not rerender on redux state change since it receives no props from the redux state.

I also second the change from a boolean to string. It is more descriptive and easier to extend later if we feel the need for more node modes.

Reply:
Yes this seems the best solution. Currently I have implemented the above solution with additional caching of edges and vertices so that no redraw occurs. Will change later this to the solution with redux state change.